OpenEdXMobile/src/main/java/org/edx/mobile/authentication/LoginAPI.java

```diff
@@ -178,15 +183,27 @@ private void register(Bundle parameters) throws Exception {
         for (String key : parameters.keySet()) {
             parameterMap.put(key, parameters.getString(key));
         }
-        Response<ResponseBody> response = loginService.register(parameterMap).execute();
+        final Response<ResponseBody> response = loginService.register(parameterMap).execute();
         if (!response.isSuccessful()) {
             final int errorCode = response.code();
             final String errorBody = response.errorBody().string();
             if ((errorCode == HttpStatus.BAD_REQUEST || errorCode == HttpStatus.CONFLICT) && !android.text.TextUtils.isEmpty(errorBody)) {
                 try {
-                    final FormFieldMessageBody body = gson.fromJson(errorBody, FormFieldMessageBody.class);
-                    if (body != null && body.size() > 0) {
-                        throw new RegistrationException(body);
+                    final Type objectMapType = new TypeToken<HashMap<String, Object>>() {
+                    }.getType();
+                    final Type fieldErrorListType = new TypeToken<List<RegisterResponseFieldError>>() {
+                    }.getType();
+                    final HashMap<String, Object> body = gson.fromJson(errorBody, objectMapType);
+                    final FormFieldMessageBody errorResponse = new FormFieldMessageBody();
+                    for (String key : body.keySet()) {
+                        final Object fieldError = body.get(key);
+                        if (fieldError instanceof Collection<?>) {
+                            // convert the object into json string otherwise throw 'com.google.gson.stream.MalformedJsonException'
+                            errorResponse.put(key, gson.fromJson(gson.toJson(fieldError), fieldErrorListType));
+                        }
```

Comment:
The following code would be more descriptive and understandable:

```java
final Object fieldError = body.get(key);
if (fieldError instanceof Collection<?>) {
    /*
    The conversion of the errorBody from JSON String to Map<String, Object>
    malformed the JSON properties of key & value pairs inside it i.e.
    removed the quotations encapsulating the key & value pairs.

    Following code line is required to transform them back into valid JSON.
     */
    final String fieldErrorJson = gson.toJson(fieldError);
    errorResponse.put(key, gson.fromJson(fieldErrorJson, fieldErrorListType));
}
```

Reply:
Done.